### ISSUE TITLE
Problem with amazonaws.com

### ISSUE THREAD

**Issue by UnveilTech** [usrflo/registered-domain-libs#16]
Hi,
is that normal that none of amazonaws.com hosts are found ?
ex: kinesis.us-east-1.amazonaws.com
each time it gives NULL

bye Fred

**Comment by Pascal76** [comment]
same issue (s3.amazonaws.com) :(

+ Deprecated: strlen(): Passing null to parameter #1 ($string) of type string is deprecated in <...>/registered-domain-libs/regDomain.class.php on line 69

(on php 8.1)

**Comment by usrflo** [comment]
@UnveilTech : I tried all 3 implementations with `kinesis.us-east-1.amazonaws.com` and the results look good to me:

```
registered-domain-libs/PHP $ php test-regDomain.php kinesis.us-east-1.amazonaws.com
kinesis.us-east-1.amazonaws.com
registered-domain-libs/PHP $ php test-regDomain.php test.kinesis.us-east-1.amazonaws.com
kinesis.us-east-1.amazonaws.com

registered-domain-libs/Perl $ perl -I. test-regDomain.pl kinesis.us-east-1.amazonaws.com
kinesis.us-east-1.amazonaws.com
registered-domain-libs/Perl $ perl -I. test-regDomain.pl test.kinesis.us-east-1.amazonaws.com
kinesis.us-east-1.amazonaws.com

registered-domain-libs/C $ ./test-regdom kinesis.us-east-1.amazonaws.com
kinesis.us-east-1.amazonaws.com: kinesis.us-east-1.amazonaws.com
registered-domain-libs/C $ ./test-regdom test.kinesis.us-east-1.amazonaws.com
test.kinesis.us-east-1.amazonaws.com: kinesis.us-east-1.amazonaws.com
```
What exactly fails with NULL?

**Comment by usrflo** [comment]
> same issue (s3.amazonaws.com) :(

@Pascal76: as `s3.amazonaws.com` is listed in https://publicsuffix.org/list/public_suffix_list.dat it is considered to be a public suffix according to the definition at https://publicsuffix.org/ : _'A "public suffix" is one under which Internet users can (or historically could) directly register names'_.

This regdom-lib allows to detect registered domain names equal to the subdomain below a public suffix.
Because of the list record `s3.amazonaws.com` you get as a result:
* s3.amazonaws.com is no registered domain
* sub.s3.amazonaws.com could be a registered domain

Looking at the request of an Amazon employee in https://github.com/publicsuffix/list/issues/105 to add `s3.amazonaws.com` to the public suffix list I think Amazon may have had a different requirement: as you can see at https://publicsuffix.org/learn/ in section "Uses" browsers e.g. restrict cookie settings with these public suffixes. It may be Amazon tried to force some restrictions with their detailed list of suffixes. And this may break the detection of the registered domain that in my view should be `amazonaws.com` in case of `s3.amazonaws.com`. Let's see if they can clarify this point.

**Comment by Pascal76** [comment]
` public function validDomainPart($domPart) {
                // see http://www.register.com/domain-extension-rules.rcmx
                $len = strlen($domPart);
`   
There should be a check to avoid NULL for $domPart

**Comment by usrflo** [comment]
> There should be a check to avoid NULL for $domPart

I just added it in https://github.com/usrflo/registered-domain-libs/commit/ee64e44b0f127892b433e7de4985b7e88179ee80

**Comment by UnveilTech** [comment]
Hi,
<?php
$szHost = "kinesis.us-east-1.amazonaws.com";
$szDom = getRegisteredDomain($szHost);
?>

$szDom is NULL, it should be "amazonaws.com"

bye Fred

**Comment by UnveilTech** [comment]
$szHost = "kinesis.us-east-1.amazonaws.com";
$szDom = getRegisteredDomain($szHost);

**Comment by usrflo** [comment]
> $szDom is NULL, it should be "amazonaws.com"

$szDom is "kinesis.us-east-1.amazonaws.com" according to the explanation in https://github.com/usrflo/registered-domain-libs/issues/16#issuecomment-1418247938.

Maybe it would be more precise to differ between
a) registered domain like amazonaws.com or dyndns.org
b) domain that can be used to restrict cookie settings for. Otherwise sites at *.amazonaws.com or *.dyndns.org could all collide/access cookies at amazonaws.com/dyndns.org

What's your use-case for getRegisteredDomain(...) ?

**Comment by UnveilTech** [comment]
What's your use-case for getRegisteredDomain(...) ?

$szDom = getRegisteredDomain("kinesis.us-east-1.amazonaws.com");
result is NULL for $szDom

**Comment by usrflo** [comment]
Are you sure you're using the latest implementation?

$registeredDomain = getRegisteredDomain("kinesis.us-east-1.amazonaws.com", $tldTree);
--> the result in $registeredDomain is "kinesis.us-east-1.amazonaws.com"

**Comment by UnveilTech** [comment]
Hi,
latest implementation does not return NULL anymore, that's good.
but it should return the master domain "amazonaws.com" instead of "kinesis.us-east-1.amazonaws.com", no ?

bye Fred

**Comment by usrflo** [comment]
> but it should return the master domain "amazonaws.com" instead of "kinesis.us-east-1.amazonaws.com", no ?

@UnveilTech : yes, sure. Can you confirm that my explanation for this issue in https://github.com/usrflo/registered-domain-libs/issues/16#issuecomment-1418247938 makes the problem clear?
